src/jit/hwintrinsicxarch.cpp

```diff
@@ -895,7 +895,44 @@ GenTree* Compiler::impSSE41Intrinsic(NamedIntrinsic        intrinsic,
                                      CORINFO_SIG_INFO*     sig,
                                      bool                  mustExpand)
 {
-    return nullptr;
+    GenTree* retNode  = nullptr;
+    GenTree* op1      = nullptr;
+    GenTree* op2      = nullptr;
+    GenTree* op3      = nullptr;
+    GenTree* op4      = nullptr;
+    int      simdSize = simdSizeOfHWIntrinsic(intrinsic, sig);
+
+    assert(simdSize == 16);
+
+    switch (intrinsic)
+    {
+        case NI_SSE41_CeilingScalar:
+        case NI_SSE41_FloorScalar:
+        case NI_SSE41_RoundCurrentDirectionScalar:
+        case NI_SSE41_RoundToNearestIntegerScalar:
+        case NI_SSE41_RoundToNegativeInfinityScalar:
+        case NI_SSE41_RoundToPositiveInfinityScalar:
+        case NI_SSE41_RoundToZeroScalar:
+        {
+            assert((sig->numArgs == 1) || (sig->numArgs == 2));
+            var_types baseType = getBaseTypeOfSIMDType(sig->retTypeSigClass);
+            assert((baseType == TYP_FLOAT) || (baseType == TYP_DOUBLE));
+
+            if (sig->numArgs == 2)
+            {
+                op2 = impSIMDPopStack(TYP_SIMD16);
+            }
+
+            op1     = impSIMDPopStack(TYP_SIMD16);
+            retNode = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, op2, intrinsic, baseType, simdSize);
```

Comment:
For `CeilingScalar/FloorScalar`, I think we can duplicate the source-operand for one-arg overloads in the importer. Then we can set `NumArg` as `2` in the table, which may simplify the implementation.

Reply:
I suggest we do that, in general, later and only after we have confirmed the codegen will be as good.

@CarolEidt, do you know if the JIT is smart enough to recognize that `op2` is a clone of `op1` and only needs to be read from memory once?

Reply:
> I suggest we do that, in general, later and only after we have confirmed the codegen will be as good.

It is okay to me. But `numArgOfHWIntrinsic` should be modified if we distinguish one-arg and two-arg overloads in the back-end.

Reply:
> op2 is a clone of op1 and only needs to be read from memory once

Not sure what "clone" means here. But the "only read from memory once" typically happens only if CSE kicks in.

Reply:
> Not sure what "clone" means here

@mikedn, `gtCloneExpr(op1)`.

Currently, we only have `op1` (`op2` is nullptr) and it's all we carry through codegen. `op1` will get loaded into register (via `movups`) and we will ultimately emit `INS op1Reg, op1Reg, imm8`.

If we instead do what @fiigii is suggesting my concern would be that the JIT isn't smart enough to always recognize that op1 and op2 represent the same underlying location. If it isn't smart enough we might end up emitting two separate `movups` and `INS op1Reg, op2Reg, imm8` or one `movups` and a folded load (`INS [op1], op2Reg, imm8`).

Reply:
If you do end up cloning, `impCloneExpr` may be what you're looking for, it clones "cheap" trees and creates new temps for "expensive" trees.

Reply:
I would agree that keeping as-is is best, at least until such time as we come across a good reason for doing otherwise.